executor/infoschema_reader.go

```diff
@@ -201,13 +201,11 @@ func (c *statsCache) setLoading(loading bool) {
 
 func (c *statsCache) get(ctx sessionctx.Context) (map[int64]uint64, map[tableHistID]uint64, error) {
 	c.mu.Lock()
-	if time.Since(c.modifyTime) < TableStatsCacheExpiry || c.loading {
+	if time.Since(c.modifyTime) < TableStatsCacheExpiry && c.loading {
 		tableRows, colLength := c.tableRows, c.colLength
 		c.mu.Unlock()
 		return tableRows, colLength, nil
 	}
-	c.loading = true
-	c.mu.Unlock()
 
 	tableRows, err := getRowCountAllTable(ctx)
 	if err != nil {
```

Comment:
if this `err` happens, it will deadlock.
Prefer to use `defer` in this situation.
```
c.mu.Lock()
defer c.mu.Unlock()
```

Reply:
great!

Reply:
For best performance, we can use `sync.RWMutex` here.
```go
// fast path
c.mu.RLock()
if time.Since(c.modifyTime) < TableStatsCacheExpiry {
    tableRows, colLength := c.tableRows, c.colLength
    c.mu.RUnlock()
    return tableRows, colLength, nil
}
c.mu.RUnlock()


c.mu.Lock()
defer c.mu.Unlock()
if time.Since(c.modifyTime) < TableStatsCacheExpiry {
    return c.tableRows, c.colLength, nil
}
// slow path: fetch from remote

```

Reply:
Great！done.